Support `defaultValue` as single value for multi-lang bones

Currently you must provide the `defaultValue` in the structure of the internal bone value:
```py
test_boolean_single_lang1 = BooleanBone(
    descr="test_boolean_single_lang1",
    languages=["de", "en", "fr"],
    defaultValue={
        "de": True,
        "en": None,
        "fr": False,
    },
)
```
But in case you want wo initialize all languages with the same value, it would be nice to write just `defaultValue=True` instead of `defaultValue={lang: True, ...}`.


---

Bug 2 is still a problem with a dict:
```py
test_boolean_single_lang1 = BooleanBone(
    descr="test_boolean_single_lang1",
    languages=["de", "en", "fr"],
    defaultValue={
        "de": False,
        "en": False,
        "fr": False,
    },
)
```
because of this (`and boneVal[language]`) falsy test: https://github.com/viur-framework/viur-core/blob/d9cea574abbfa689551e825d42c8e8a21c59b0ab/src/viur/core/render/json/default.py#L114


 

> ### Bug 1
> 
> This bone:
> ```py
> test_boolean_single_lang1 = BooleanBone(
>     descr="test_boolean_single_lang1",
>     languages=["de", "en", "fr"],
>     defaultValue=True,
> )
> ```
> 
> causes this error:
> ```py
> [2024-03-08 17:36:24,205] viur/core/request.py:331 [ERROR] argument of type 'bool' is not iterable
> Traceback (most recent call last):
>   File "/project/deploy/viur/core/request.py", line 305, in _process
>     self._route(path)
>   File "/project/deploy/viur/core/request.py", line 561, in _route
>     res = caller(*self.args, **kwargs)
>   File "/project/deploy/viur/core/module.py", line 299, in __call__
>     return self._func(self._instance, *args, **kwargs)
>   File "/project/deploy/viur/core/prototypes/list.py", line 272, in add
>     return self.render.add(skel)
>   File "/project/deploy/viur/core/render/json/default.py", line 219, in add
>     return self.renderEntry(skel, action, params)
>   File "/project/deploy/viur/core/render/json/default.py", line 162, in renderEntry
>     vals = self.renderSkelValues(skel)
>   File "/project/deploy/viur/core/render/json/default.py", line 137, in renderSkelValues
>     res[key] = self.renderBoneValue(bone, skel, key)
>   File "/project/deploy/viur/core/render/json/default.py", line 113, in renderBoneValue
>     if boneVal and language in boneVal and boneVal[language]:
> TypeError: argument of type 'bool' is not iterable
> ```
> 
> 
> ---
> 
> ### Bug 2
> 
> These bones:
> ```py
> test_boolean_single1 = BooleanBone(
>     descr="test_boolean_single1",
>     defaultValue=True,
> )
> test_boolean_single2 = BooleanBone(
>     descr="test_boolean_single2",
>     defaultValue=False,
> )
> test_boolean_single3 = BooleanBone(
>     descr="test_boolean_single3",
>     defaultValue=None,
> )
> # test_boolean_single_lang1 = BooleanBone(
> #     descr="test_boolean_single_lang1",
> #     languages=["de", "en", "fr"],
> #     defaultValue=True,
> # )
> test_boolean_single_lang2 = BooleanBone(
>     descr="test_boolean_single_lang2",
>     languages=["de", "en", "fr"],
>     defaultValue=False,
> )
> test_boolean_single_lang3 = BooleanBone(
>     descr="test_boolean_single_lang3",
>     languages=["de", "en", "fr"],
>     defaultValue=None,
> )
> ```
> 
> caused this part of value response on `/add`:
> ```json
> "test_boolean_single1": true,
> "test_boolean_single2": false,
> "test_boolean_single3": null,
> "test_boolean_single_lang2": {
>   "de": null,
>   "en": null,
>   "fr": null
> },
> "test_boolean_single_lang3": {
>   "de": null,
>   "en": null,
>   "fr": null
> }
> ```
> 
> As you can see the values in `test_boolean_single_lang2` are `null` instead of `false`.

_Originally posted by @sveneberth in https://github.com/viur-framework/viur-core/pull/988#pullrequestreview-1925308461_

--- a/src/viur/core/bones/base.py
+++ b/src/viur/core/bones/base.py
@@ -228,6 +228,8 @@ class BaseBone(object):
              and all([isinstance(x, str) for x in languages]))
         ):
             raise ValueError("languages must be None or a list of strings")
+        if languages and "__default__" in languages:
+            raise ValueError("__default__ is not supported as a language")
         if (
             not isinstance(required, bool)
             and (not isinstance(required, (tuple, list)) or any(not isinstance(value, str) for value in required))
@@ -244,8 +246,16 @@ class BaseBone(object):
         # Default value
         # Convert a None default-value to the empty container that's expected if the bone is
         # multiple or has languages
-        if defaultValue is None and self.languages:
-            self.defaultValue = {}
+        if self.languages:
+            if not isinstance(defaultValue, dict):
+                self.defaultValue = {lang: defaultValue for lang in self.languages}
+            else:
+                if "__default__" in defaultValue:
+                    self.defaultValue = {lang: defaultValue["__default__"] for lang in self.languages}
+                    self.defaultValue |= {lang: defaultValue[lang] for lang in self.languages if lang in defaultValue}
+                else:
+                    self.defaultValue = defaultValue
+
         elif defaultValue is None and self.multiple:
             self.defaultValue = []
         else:

--- a/src/viur/core/bones/base.py
+++ b/src/viur/core/bones/base.py
@@ -251,8 +251,7 @@ class BaseBone(object):
                 self.defaultValue = {lang: defaultValue for lang in self.languages}
             else:
                 if "__default__" in defaultValue:
-                    self.defaultValue = {lang: defaultValue["__default__"] for lang in self.languages}
-                    self.defaultValue |= {lang: defaultValue[lang] for lang in self.languages if lang in defaultValue}
+                    self.defaultValue = {lang: defaultValue.get(lang, defaultValue["__default__"]) for lang in self.languages}
                 else:
                     self.defaultValue = defaultValue
 

--- a/src/viur/core/bones/base.py
+++ b/src/viur/core/bones/base.py
@@ -249,11 +249,11 @@ class BaseBone(object):
         if self.languages:
             if not isinstance(defaultValue, dict):
                 self.defaultValue = {lang: defaultValue for lang in self.languages}
+            elif "__default__" in defaultValue:
+                self.defaultValue = {lang: defaultValue.get(lang, defaultValue["__default__"])
+                                     for lang in self.languages}
             else:
-                if "__default__" in defaultValue:
-                    self.defaultValue = {lang: defaultValue.get(lang, defaultValue["__default__"]) for lang in self.languages}
-                else:
-                    self.defaultValue = defaultValue
+                self.defaultValue = defaultValue
 
         elif defaultValue is None and self.multiple:
             self.defaultValue = []

